File: test/parsergenerator/closureitem_test.scala
package pg.parsergenerator

import org.scalatest.FunSpec
import pg.parsergenerator.closureitem.{genClosureItem,isSameLR0, isSameLR1}
import pg.data.sample_language.{test_sample_language}
import pg.parsergenerator.grammardb.{genGrammarDB,getTokenId}
import pg.token.{SYMBOL_EOF}

class closureitem_test extends FunSpec {

  describe("ClosureItem test") {
    val grammardb = genGrammarDB(test_sample_language)
    describe("{S' -> . S [$]}") {
      val ci = genClosureItem(grammardb, -1, 0, Array(SYMBOL_EOF))
      it("getter") {
        assert(ci.rule_id == -1)
        assert(ci.dot_index == 0)
        assertResult(ci.lookaheads )( Array(SYMBOL_EOF))
      }
      it("ClosureItem Hash") {
        val id_eof = getTokenId(grammardb, SYMBOL_EOF)
        assert(ci.lr0_hash == "-1,0")
        assert(ci.lr1_hash == "-1,0,["+id_eof+"]")
      }
      describe("ClosureItem equality") {
        it("compare itself") {
          assert(isSameLR0(ci, ci))
          assert(isSameLR1(ci, ci))
        }
        it("same ClosureItem") {
          val ci2 = genClosureItem(grammardb, -1, 0, Array(SYMBOL_EOF))
          assert(isSameLR0(ci, ci2))
          assert(isSameLR1(ci, ci2))
        }
        it("not same ClosureItem") {
          val ci2 = genClosureItem(grammardb, 0, 0, Array(SYMBOL_EOF))
          assert(!isSameLR0(ci, ci2))
          assert(!isSameLR1(ci, ci2))
        }
        it("not same lookahead item") {
          val ci2 = genClosureItem(grammardb, -1, 0, Array("ID"))
          assert(isSameLR0(ci, ci2))
          assert(!isSameLR1(ci, ci2))
        }
      }
      it("invalid lookahead item") {
        assertThrows[Error](genClosureItem(grammardb, -1, 0, Array("X"))) // invalid token
      }
    }
    describe("invalid ClosureItem") {
      it("invalid grammar id") {
        assertThrows[Error](genClosureItem(grammardb, -2, 0, Array(SYMBOL_EOF)))
      }
      it("invalid dot position") {
        assertThrows[Error](genClosureItem(grammardb, -1, -1, Array(SYMBOL_EOF)))
      }
    }
  }
}

File: test/yacc/closureitem_test.ml
open OUnit
open Closureitem
open Grammardb
open Token
open Language

let test () =
  let grammardb = genGrammarDB(Sample_language.language) in
  let ci = genClosureItem grammardb (-1) 0 ["EOF"] in
  "ClosureItem test" >::: [
    "{S' -> . S [$]}" >::: [
      "getter" >:: begin fun () ->
        assert(ci.rule_id = -1);
        assert(ci.dot_index = 0);
        assert_equal (ci.lookaheads) ["EOF"]
      end;
      "ClosureItem Hash" >:: begin fun () ->
        let id_eof = getTokenId grammardb "EOF" in
        assert(ci.lr0_hash = "-1,0");
        assert(ci.lr1_hash = "-1,0,["^string_of_int id_eof^"]")
      end;
      "ClosureItem equality" >::: [
        "compare itself" >:: begin fun () ->
          assert(isSameLR0 ci ci);
          assert(isSameLR1 ci ci)
        end;
        "same ClosureItem" >:: begin fun () ->
          let ci2 = genClosureItem grammardb (-1) 0 ["EOF"] in
          assert(isSameLR0 ci ci2);
          assert(isSameLR1 ci ci2)
        end;
        "not same ClosureItem" >:: begin fun () ->
          let ci2 = genClosureItem grammardb 0 0 ["EOF"] in
          assert (not(isSameLR0 ci ci2));
          assert (not(isSameLR1 ci ci2))
        end;
        "not same lookahead item" >:: begin fun () ->
          let ci2 = genClosureItem grammardb (-1) 0 ["ID"] in
          assert(isSameLR0 ci ci2);
          assert(not(isSameLR1 ci ci2))
        end;
      ];
      "invalid lookahead item" >:: begin fun () ->
        assert_raises (Failure "invalid token X")
          (fun () -> genClosureItem grammardb (-1) 0 ["X"])
      end;
    ];
    "invalid ClosureItem" >::: [
      "invalid grammar id" >:: begin fun () ->
        assert_raises (Failure "invalid grammar id")
          (fun () -> genClosureItem grammardb (-2) 0 ["EOF"])
      end;
      "invalid dot position" >:: begin fun () ->
        assert_raises (Failure "dot index out of range")
          (fun () -> genClosureItem grammardb (-1) (-1) ["EOF"])
      end;
    ];
  ]

